internal/gps/result.go

```diff
@@ -65,37 +66,91 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 		return err
 	}
 
-	var wg sync.WaitGroup
-	errCh := make(chan error, len(l.Projects()))
+	lps := l.Projects()
+
+	type resp struct {
+		i   int
+		err error
+	}
+	respCh := make(chan resp, len(lps))
+	writeCh := make(chan int, len(lps))
+	cancel := make(chan struct{})
 
-	for _, p := range l.Projects() {
-		wg.Add(1)
-		go func(p LockedProject) {
+	// Queue work.
+	for i := range lps {
+		writeCh <- i
+	}
+	close(writeCh)
+	// Launch writers.
+	writers := runtime.GOMAXPROCS(-1)
+	if len(lps) < writers {
+		writers = len(lps)
+	}
+	var wg sync.WaitGroup
+	wg.Add(writers)
+	for i := 0; i < writers; i++ {
+		go func() {
 			defer wg.Done()
-			to := filepath.FromSlash(filepath.Join(basedir, string(p.Ident().ProjectRoot)))
-			logger.Printf("Writing out %s@%s", p.Ident().errString(), p.Version())
 
-			if err := sm.ExportProject(p.Ident(), p.Version(), to); err != nil {
-				errCh <- errors.Wrapf(err, "failed to export %s", p.Ident().ProjectRoot)
-				return
-			}
+			for i := range writeCh {
+				select {
+				case <-cancel:
+					return
+				default:
+				}
 
-			if sv {
-				err := filepath.Walk(to, stripVendor)
-				if err != nil {
-					errCh <- errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)
+				p := lps[i]
+				to := filepath.FromSlash(filepath.Join(basedir, string(p.Ident().ProjectRoot)))
+
+				if err := sm.ExportProject(p.Ident(), p.Version(), to); err != nil {
+					respCh <- resp{i, errors.Wrapf(err, "failed to export %s", p.Ident().ProjectRoot)}
+					continue
+				}
+
+				if sv {
+					select {
+					case <-cancel:
+						return
+					default:
+					}
+
+					if err := filepath.Walk(to, stripVendor); err != nil {
+						respCh <- resp{i, errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)}
+						continue
+					}
 				}
+
+				respCh <- resp{i, nil}
 			}
-		}(p)
+		}()
+	}
+	// Monitor writers
+	go func() {
+		wg.Wait()
+		close(respCh)
+	}()
+
+	// Log results and collect errors
+	var errs []error
+	var cnt int
+	for resp := range respCh {
+		cnt++
+		msg := "Wrote"
+		if resp.err != nil {
+			if len(errs) == 0 {
+				close(cancel)
```

Comment:
need to know to break out of the loop explicitly here. i believe there's a possibility of a double-close on the cancel channel:

1. X and Y are all sent into the `writeCh`, and are all picked up by their own goroutine
2. X finishes first and fails, indicates as much to the `respCh`; it terminates, decrementing the wg
3. main goroutine receives the fail and closes `cancel`, logs result, returns to loop and waits for new value
4. meanwhile, Y finishes and also fails, sends to `respCh`, and terminates, closing `respCh` from the waiter goroutine
5. however, main goroutine is still waiting on the respCh and is already working before it receives second fail, panics on double-close

note that because the `range` loop will not terminate on a buffered channel until the chan is both closed AND empty, i believe this is guaranteed to happen if two workers simultaneously encounter an error.

Reply:
The `close(cancel)` is protected by `len(errs) == 0` so that the channel is only closed on the first error.  We could break and abort on the first error (and only report that single error), but since we want to gracefully block/wait/shutdown anyways, it seemed cleaner to inspect all response and collect all errors (it might be relevant to investigating the error; we don't want to block/wait in silence then not log anything; etc...).

Reply:
gah, you're right, i just breezed right by that check. ok 👍